palace/models/romoperator.cpp

```diff
@@ -77,9 +108,86 @@ inline void ProjectVecInternal(MPI_Comm comm, const std::vector<Vector> &V,
   Mpi::GlobalSum(n - n0, br.data() + n0, comm);
 }
 
+inline void ComputeMRI(const Eigen::MatrixXcd &R, Eigen::VectorXcd &q)
+{
+  // Compute the coefficients of the minimal rational interpolation (MRI):
+  // u = [sum_i u_i q_i / (z - z_i)] / [sum_i q_i / (z - z_i)]. The coefficients are given
+  // by the right singular vector of R corresponding to the minimum singular value.
+  const auto S = R.rows();
+  MFEM_ASSERT(S > 0 && R.cols() == S, "Invalid dimension mismatch when computing MRI!");
+  // For Eigen = v3.4.0 (latest tagged release as of 10/2023)
+  Eigen::JacobiSVD<Eigen::MatrixXcd> svd;
+  svd.compute(R, Eigen::ComputeFullV);
+  // For Eigen > v3.4.0 (GitLab repo is at v3.4.90 as of 10/2023)
+  // Eigen::JacobiSVD<Eigen::MatrixXcd, Eigen::ComputeFullV> svd;
+  // svd.compute(R);
+  const auto &sigma = svd.singularValues();
+  auto m = S - 1;
+  while (m > 0 && sigma[m] < ORTHOG_TOL * sigma[0])
+  {
+    Mpi::Warning("Minimal rational interpolation encountered rank-deficient matrix: "
+                 "σ[{:d}] = {:.3e} (σ[0] = {:.3e})!\n",
+                 m, sigma[m], sigma[0]);
+    m--;
+  }
+  q = svd.matrixV().col(m);
+}
+
+template <typename MatType, typename VecType>
+inline void ZGGEV(MatType &A, MatType &B, VecType &D, MatType &VR)
+{
+  // Wrapper for LAPACK's (z)ggev. A and B are overwritten by their Schur decompositions.
+  MFEM_VERIFY(A.rows() == A.cols() && B.rows() == B.cols() && A.rows() == B.rows(),
+              "Generalized eigenvalue problem expects A, B matrices to be square and have "
+              "same dimensions!");
+  char jobvl = 'N', jobvr = 'V';
+  int n = static_cast<int>(A.rows()), lwork = 2 * n;
+  std::vector<std::complex<double>> alpha(n), beta(n), work(lwork);
+  std::vector<double> rwork(8 * n);
+  MatType VL(0, 0);
+  VR.resize(n, n);
+  int info = 0;
+
+  zggev_(&jobvl, &jobvr, &n, A.data(), &n, B.data(), &n, alpha.data(), beta.data(),
+         VL.data(), &n, VR.data(), &n, work.data(), &lwork, rwork.data(), &info);
+  MFEM_VERIFY(info == 0, "ZGGEV failed with info = " << info << "!");
+
+  // Postprocess the eigenvalues and eigenvectors (return unit 2-norm eigenvectors).
+  D.resize(n);
+  for (int i = 0; i < n; i++)
+  {
+    D(i) = (beta[i] == 0.0)
+               ? ((alpha[i] == 0.0) ? std::numeric_limits<std::complex<double>>::quiet_NaN()
+                                    : mfem::infinity())
```

Comment:
I don't think a / b with a > 0 and b = 0 is nan, it should be infinity in floating point (signed zero). So the case handling here is just explicit in case compiler flags are specified such that a divide by zero would otherwise raise an exception or something. `mfem::infinity` just wraps `numeric_limits<double>::infinity`, which I think is already going to be the standard return for a / b with b = 0 but I wasn't certain if it would be in all cases.

Reply:
Sounds good. I was thinking in mathematical terms but forgetting the floating point difference.